pkg/serviceaccount/tokens_controller.go

```diff
@@ -240,8 +244,16 @@ func (e *TokensController) secretDeleted(obj interface{}) {
 		return
 	}
 
-	if _, err := e.removeSecretReferenceIfNeeded(serviceAccount, secret.Name); err != nil {
-		glog.Error(err)
+	for i := 1; i <= NumServiceAccountRemoveReferenceRetries; i++ {
```

Comment:
I mean delete the current list; then search through all secrets and rebuild the list. Just do it every time you process the service account; it should be quick since you're using an indexing store.